test/factory.js

```diff
@@ -272,27 +300,51 @@ describe("Test Wallet Factory", function () {
+            let tx = await factory.from(infrastructure).createCounterfactualWallet(owner.address, modules, label, salt);
             let txReceipt = await factory.verboseWaitForTransaction(tx);
             let walletAddr = txReceipt.events.filter(event => event.event == 'WalletCreated')[0].args.wallet;
             // we test that the wallet is at the correct address
             assert.equal(futureAddr, walletAddr, 'should have the correct address');
             // we create the second wallet
-            await assert.revert(factory.from(infrastructure).createCounterfactualWallet(owner.address, modules, NO_ENS, salt, { gasLimit: 500000 }), "should fail when address is in use");
+            await assert.revert(factory.from(infrastructure).createCounterfactualWallet(owner.address, modules, label, salt), "should fail when address is in use");
         });
 
         it("should fail to create when there is no modules", async () => {
             let salt = bigNumberify(randomBytes(32)).toHexString ();
+            let label = "wallet" + index; 
             let modules = [];
-            await assert.revert(factory.from(deployer).createCounterfactualWallet(owner.address, modules, NO_ENS, salt, { gasLimit: 500000 }), "should fail when modules is empty");
+            await assert.revert(factory.from(deployer).createCounterfactualWallet(owner.address, modules, label, salt), "should fail when modules is empty");
+        });
+
+        it("should fail to create when there is no ENS", async () => {
+            let salt = bigNumberify(randomBytes(32)).toHexString ();
+            let label = ""; 
+            let modules = [module1.contractAddress, module2.contractAddress];
+            await assert.revert(factory.from(deployer).createCounterfactualWallet(owner.address, modules, label, salt), "should fail when there is no ENS");
+        });
+
+        it("should emit and event when the balalnce is non zero at creation", async () => {
+            let salt = bigNumberify(randomBytes(32)).toHexString (); 
+            let label = "wallet" + index; 
+            let modules = [module1.contractAddress, module2.contractAddress]; 
+            // we get the future address
+            let futureAddr = await factory.getAddressForCounterfactualWallet(owner.address, modules, salt); 
+            // We send ETH to the address
+            await infrastructure.sendTransaction({ to: futureAddr, value: ethers.utils.bigNumberify('1000000000000000000') });
+            // we create the wallet
+            let tx = await factory.from(infrastructure).createCounterfactualWallet(owner.address, modules, label, salt);
+            let txReceipt = await factory.verboseWaitForTransaction(tx);
+            let wallet = deployer.wrapDeployedContract(Wallet, futureAddr);
+            assert.isTrue(await utils.hasEvent(txReceipt, wallet, "Received"), "should have generated Received event");
```

Comment:
There are 4 different `Received` events in contracts, how do I know `BaseWallet.init()` was the one that was triggered?

Reply:
Since we are looking at the events generated by a call to `createCounterfactualWallet()` only `BaseWalet.init()` can trigger the event.

I have nonetheless added additional checks in the test to verify that the amount is correct and that the `sender` property of the event is `address(0)`.